tests/test_decompose.cpp

```diff
@@ -163,19 +163,94 @@ int test_decompose_weighted(const std::vector<problem::base_ptr> &probs, double
 
 }
 
-int main()
+//Test decompose(TCHEBYCHEFF) when weights are given.
+int test_decompose_tchebycheff(const std::vector<problem::base_ptr> &probs, std::vector<fitness_vector> &ideal, double d_from_center)
 {
 	
+	for(unsigned int i=0; i<probs.size(); i++)
+	{
+
+		decision_vector x = construct_test_point(probs[i], d_from_center);
+
+		int f_dimension = probs[i]->get_f_dimension();
+		fitness_vector weights(f_dimension, 0);
+
+		//generate random weights vector
+		generate_weights(weights);
+
+		//get the original fitness
+		fitness_vector f_original = probs[i]->objfun(x);
+
+		//calculate expected fitness
+		double f_expected = 0.0;
+		double tmp;
+		for(int j = 0; j < f_dimension; ++j)
+		{
+			tmp = weights[j] * fabs(f_original[j] - ideal[i][j]);
+			if(tmp > f_expected)
+				f_expected = tmp;
+		}
+
+		problem::decompose prob_decompose(*(probs[i]), problem::decompose::TCHEBYCHEFF, weights);
+		
+		//set ideal point
+		prob_decompose.set_ideal_point(ideal[i]);
+
+		//get the decomposed fitness
+		fitness_vector f_decompose = prob_decompose.objfun(x);
+
+		if(is_eq_vector(weights, prob_decompose.get_weights()))
+			std::cout<<prob_decompose.get_name()<<"TCHEBYCHEFF weights passes, "<<std::endl;
+		else
+		{
+			std::cout<<prob_decompose.get_name()<<"TCHEBYCHEFF weights failed, "<<std::endl;
+			PRINT_VEC(weights);
+			PRINT_VEC(prob_decompose.get_weights());
+			return 1;
+		}
+
+		if(is_eq_vector(ideal[i], prob_decompose.get_ideal_point()))
```

Comment:
I would (re)move this test (and the test in line 202). You basically test the constructor and the setter here. I would either remove it, or do it in a separate function.